redaxo/src/core/lib/extension.php

```diff
@@ -25,15 +25,17 @@ abstract class rex_extension
     /**
      * Registers an extension point.
      *
-     * @param rex_extension_point $extensionPoint Extension point
-     *
-     * @return mixed Subject, maybe adjusted by the extensions
+     * @template T
+     * @param rex_extension_point<T> $extensionPoint Extension point
+     * @return T Subject, maybe adjusted by the extensions
```

Comment:
100% korrekt ist es nicht: Hier wird jetzt davon ausgegangen, dass der Subjekt-Typ, den der EP reingibt, auch am Ende wieder rauskommt.
Theoretisch können Extension aber ja den Typen ändern. Das lässt sich aber nicht verknüpfen, die EP-Stellen mit den Extension-Stellen.

Am besten sind die eigenen EP-Klassen mit Spezifizierung für `T`.

Bin mir noch unsicher. Auch aufgrund der phpstan-Meldungen hier an verschiedenen Stellen.
Aber so fällt hier jetzt zum Beispiel ein Bug auf, siehe meldungen bzgl. `rex_backup_tar::getMessages`.

Reply:
Wahrscheinlich taugt es leider nicht so richtig, wenn ich mir die Meldungen so anschaue. :(

Reply:
> Wahrscheinlich taugt es leider nicht so richtig, wenn ich mir die Meldungen so anschaue. :(

an welcher der meldungen machst du das fest?

Reply:
An solchen wie diese:

<img width="856" alt="Bildschirmfoto 2021-03-21 um 09 11 01" src="https://user-images.githubusercontent.com/330436/111898296-6b16a000-8a25-11eb-969e-926114a1baae.png">

Dort bräuchten wir die Möglichkeit, den Typ explizit anzugeben, sodass er nicht aus `$subject` zu sehr beschränkt wird, also:
`new rex_extension_point<array>(...)`

Ich denke, das ginge mit php/psalm zurzeit aber mit extra Variable:

```php
/** @var rex_extension_point<array> $ep */
$ep = new rex_extension_point('BE_STYLE_SCSS_COMPILE', []);
$scssFiles = rex_extension::registerPoint($ep);
foreach ($scssFiles as $file) { ... }
```

Aber wenn man die Meldungen dann weiter durchschaut, gibt es auch solche wie `always true`. Das sind Stellen, wo wir die Rückgabe der Extensions explizit prüfen, und das würde ich an den Stellen nicht entfernen, gerade weil über psalm die Verknüpfung zwischen EP und Extensions nicht besteht.

Reply:
Obwohl, die angemerkten Stellen sind eigentlich doch auch wieder Stellen, wo man "bloß" den Typ aufweiten müsste.
Aber trotzdem ist mein Gedanke, ob wir uns wirklich drauf verlassen wollen, dass das verlangte zurück kommt, oder ob es nicht eigentlich eher geprüft werden sollte.

Reply:
Ich denke es macht an vielen stellen sinn zu sagen der type bleibt gleich.
Dennoch macht es auch sinn im aufrufenden code sicherzustellen was man zurückbekommt.

ich vermute es bräuchte eine ausprogrammierte extension, damit aus „empty-array“ z.b. „array“ wird.

Reply:
Ggf reicht auch aber auch ein `assert(is_array))` damit die tools von ihrem „empty“ abweichen

Reply:
Scheint zu passen

Reply:
Alternativ: in phpstan das template auf mixed typen, wenn psalm damit umgehen kann

Reply:
das template ist ja schon mixed, es hat ja keine einschränkung. die bekommt es ja erst durch das subjekt.
oder meinst du `@phpstan-return mixed`, also return ganz ohne template?

Reply:
Ja. Wenn psalm erreicht was du willst und wir phpstan so ruhig bekommen, dann finde ich spricht nix dagegen

Reply:
Eigentlich macht es phpstan schon richtig. Und ich habe jetzt mal geprüft, auch psalm erkennt es als leeres Array etc, bloß psalm meldet nicht, wenn man ein leeres Array per foreach durchläuft. Das Grundproblem besteht dort aber somit genauso.